File: Helper/Bridge/Invoker.cs
#pragma warning disable 0,
// ReSharper disable InconsistentNaming VariableHidesOuterVariable AccessToModifiedClosure

using Helper;
using Helper.Utility;

namespace Helper.Bridge {

	public static unsafe class Invoker {

		#region function

		public static String Execute (
			Library                          library,
			Func<List<String>, List<String>> callback,
			String                           script,
			List<String>                     argument
		) {
			var callbackResultHandler = default(Interface.StringList*);
			var callbackExceptionHandler = default(Interface.String*);
			var callbackProxy = Interface.String* (
				Interface.StringList** argument,
				Interface.StringList** result
			) => {
				Converter.DestructStringList(ref *callbackResultHandler);
				Converter.DestructString(ref *callbackExceptionHandler);
				*result = callbackResultHandler;
				try {
					var argumentValue = Converter.ParseStringList(ref **argument);
					var resultValue = callback(argumentValue);
					Converter.ConstructStringList(ref *callbackResultHandler, resultValue);
					return null;
				}
				catch (Exception e) {
					var exceptionValue = e.ToString();
					Converter.ConstructString(ref *callbackExceptionHandler, exceptionValue);
					return callbackExceptionHandler;
				}
			};
			var callbackProxyGuard = GCHandle.Alloc(callbackProxy);
			{
				callbackResultHandler = MemoryHelper.Allocate<Interface.StringList>();
				callbackExceptionHandler = MemoryHelper.Allocate<Interface.String>();
				Converter.ConstructStringList(ref *callbackResultHandler, new List<String>());
				Converter.ConstructString(ref *callbackExceptionHandler, "");
			}
			var callbackPointer = default(Interface.Callback*);
			var scriptPointer = default(Interface.String*);
			var argumentPointer = default(Interface.StringList*);
			var resultPointer = default(Interface.String*);
			{
				callbackPointer = MemoryHelper.Allocate<Interface.Callback>();
				scriptPointer = MemoryHelper.Allocate<Interface.String>();
				argumentPointer = MemoryHelper.Allocate<Interface.StringList>();
				Converter.ConstructCallback(ref *callbackPointer, (delegate* <Interface.String*, Interface.StringList**, Interface.StringList**>)Marshal.GetFunctionPointerForDelegate(callbackProxy));
				Converter.ConstructString(ref *scriptPointer, script);
				Converter.ConstructStringList(ref *argumentPointer, argument);
			}
			var exceptionPointer = library.Execute(&callbackPointer, &scriptPointer, &argumentPointer, &resultPointer);
			{
				Converter.DestructCallback(ref *callbackPointer);
				Converter.DestructString(ref *scriptPointer);
				Converter.DestructStringList(ref *argumentPointer);
				MemoryHelper.Free(callbackPointer);
				MemoryHelper.Free(scriptPointer);
				MemoryHelper.Free(argumentPointer);
			}
			var result = Converter.ParseString(ref *resultPointer);
			{
				Converter.DestructStringList(ref *callbackResultHandler);
				Converter.DestructString(ref *callbackExceptionHandler);
				MemoryHelper.Free(callbackResultHandler);
				MemoryHelper.Free(callbackExceptionHandler);
			}
			callbackProxyGuard.Free();
			if (exceptionPointer is not null) {
				var exception = Converter.ParseString(ref *exceptionPointer);
				throw new Exception(exception);
			}
			return result;
		}

		#endregion

	}

}

File: ShellGUI/lib/bridge/invoker.dart
import '/common.dart';
import '/bridge/interface.dart' as Interface; // ignore: library_prefixes
import '/bridge/converter.dart';
import '/bridge/library.dart';
import 'dart:ffi' as ffi;
import 'package:ffi/ffi.dart' as ffi;

// ----------------

class Invoker {

  static List<String> Function(List<String>)? _callbackImplement = null;
  static ffi.Pointer<Interface.StringList>?   _callbackResultHandler = null;
  static ffi.Pointer<Interface.String>?       _callbackExceptionHandler = null;

  static ffi.Pointer<Interface.String> _callbackProxy(
    ffi.Pointer<ffi.Pointer<Interface.StringList>> argument,
    ffi.Pointer<ffi.Pointer<Interface.StringList>> result,
  ) {
    Converter.destructStringList(_callbackResultHandler!.ref);
    Converter.destructString(_callbackExceptionHandler!.ref);
    result.value = _callbackResultHandler!;
    try {
      var argumentValue = Converter.parseStringList(argument.value.ref);
      var resultValue = _callbackImplement!(argumentValue);
      Converter.constructStringList(_callbackResultHandler!.ref, resultValue);
      return ffi.nullptr;
    }
    catch (e) {
      var exceptionValue = e.toString();
      Converter.constructString(_callbackExceptionHandler!.ref, exceptionValue);
      return _callbackExceptionHandler!;
    }
  }

  // ----------------

  static String execute(
    Library                             library,
    List<String> Function(List<String>) callback,
    String                              script,
    List<String>                        argument,
  ) {
    {
      _callbackImplement = callback;
      _callbackResultHandler = ffi.calloc.call<Interface.StringList>();
      _callbackExceptionHandler = ffi.calloc.call<Interface.String>();
      Converter.constructStringList(_callbackResultHandler!.ref, []);
      Converter.constructString(_callbackExceptionHandler!.ref, '');
    }
    var callbackPointer = ffi.Pointer<ffi.Pointer<Interface.Callback>>.fromAddress(0);
    var scriptPointer = ffi.Pointer<ffi.Pointer<Interface.String>>.fromAddress(0);
    var argumentPointer = ffi.Pointer<ffi.Pointer<Interface.StringList>>.fromAddress(0);
    var resultPointer = ffi.Pointer<ffi.Pointer<Interface.String>>.fromAddress(0);
    {
      callbackPointer = ffi.calloc.call<ffi.Pointer<Interface.Callback>>();
      scriptPointer = ffi.calloc.call<ffi.Pointer<Interface.String>>();
      argumentPointer = ffi.calloc.call<ffi.Pointer<Interface.StringList>>();
      resultPointer = ffi.calloc.call<ffi.Pointer<Interface.String>>();
    }
    {
      callbackPointer.value = ffi.calloc.call<Interface.Callback>();
      scriptPointer.value = ffi.calloc.call<Interface.String>();
      argumentPointer.value = ffi.calloc.call<Interface.StringList>();
      Converter.constructCallback(callbackPointer.value.ref, ffi.Pointer.fromFunction(_callbackProxy));
      Converter.constructString(scriptPointer.value.ref, script);
      Converter.constructStringList(argumentPointer.value.ref, argument);
    }
    var exceptionPointer = library.execute(callbackPointer, scriptPointer, argumentPointer, resultPointer);
    {
      Converter.destructCallback(callbackPointer.value.ref);
      Converter.destructString(scriptPointer.value.ref);
      Converter.destructStringList(argumentPointer.value.ref);
      ffi.calloc.free(callbackPointer.value);
      ffi.calloc.free(scriptPointer.value);
      ffi.calloc.free(argumentPointer.value);
    }
    var result = Converter.parseString(resultPointer.value.ref);
    {
      ffi.calloc.free(callbackPointer);
      ffi.calloc.free(scriptPointer);
      ffi.calloc.free(argumentPointer);
      ffi.calloc.free(resultPointer);
    }
    {
      Converter.destructStringList(_callbackResultHandler!.ref);
      Converter.destructString(_callbackExceptionHandler!.ref);
      ffi.calloc.free(_callbackResultHandler!);
      ffi.calloc.free(_callbackExceptionHandler!);
      _callbackImplement = null;
      _callbackResultHandler = null;
      _callbackExceptionHandler = null;
    }
    if (exceptionPointer != ffi.nullptr) {
      var exception = Converter.parseString(exceptionPointer.ref);
      throw exception;
    }
    return result;
  }

}
